plugins/Compressor/Compressor.cpp

```diff
@@ -415,11 +356,22 @@ bool CompressorEffect::processAudioBuffer(sampleFrame* buf, const fpp_t frames)
 
 			// Keep it above the noise floor
 			m_yL[i] = qMax(COMP_NOISE_FLOOR, m_yL[i]);
+			
+			float scVal = m_yL[i];
+			
+			if (lookahead)
+			{
+				const float temp = scVal;
+				// Lookahead is calculated by picking the largest value between
+				// the current sidechain signal and the delayed sidechain signal.
+				scVal = std::max(m_scLookBuf[i][m_lookWrite], m_scLookBuf[i][(m_lookWrite + m_lookBufLength - m_lookaheadLength) % m_lookBufLength]);
```

Comment:
This algorithm looks too simple, could you explain your choice here?

Reply:
In a compressor, the purpose of lookahead is to use a short amount of latency to allow a compressor to "look into the future" to be able to react to transients before they arrive.  Without lookahead, transients will always play before the compressor knows they're coming, which means even with a very fast attack time, the initial click from them will pop through the results, and cannot be tamed in any way.  Lookahead allows the compressor to begin the attack behavior slightly before the transient actually arrives, removing this issue entirely, and giving the user a method of controlling precisely how long the compressor has to react before the transient's arrival.

Of course, simply shifting the sidechain signal backwards isn't the solution, since that would shift the release phase as well.  With even a tiny amount of lookahead, the gain envelope would lose almost all relevance to the actual audio being passed through it, and would no longer even function as compression.  We need a method of making the attack phase occur earlier, while leaving the release phase exactly where it was at.

With a maximum possible lookahead of only 20 milliseconds, an elegantly simple and incredibly effective method of approaching this is to take the maximum value between the sidechain signal and the sidechain signal with negative delay, and then using that as the actual sidechain signal.  This means the compressor will be perfectly aware of transients as if they occurred earlier by the exact amount of time set by the user via the lookahead, and meanwhile the release stage remains entirely unimpacted regardless of the lookahead length in use.  This flawlessly fixes the original issue while making only minimal changes to the compressor's actual behavior.

Here's an example of the Compressor's reaction to instant DC offset with and without lookahead.  On the left there's no lookahead, so as you can see, there's quite a large volume spike at the initial transient, because it takes the Compressor some time to fully react to the transient in the way dictated by the attack time (which I have set to only 5 milliseconds).  On the right I have the lookahead set to its maximum value, which allows the Compressor to make the proper gain reductions slightly before the transient actually occurs.  The precise length of time it has to perform this task is dictated by the lookahead knob's value.
![image](https://github.com/LMMS/lmms/assets/34612565/e2af8b84-02e4-4b8b-a7d6-e868c2347dfe)

I didn't bother providing an image to show the release phase because it's exactly the same regardless of the setting, exactly as it should be.  So, lookahead is simply a parameter to control how many milliseconds prior to each transient that the attack phase is allowed to begin.